### PR TITLE
Haptic feedback

### PR DESCRIPTION
**Pull request by Laserlicht**
Haptic feedback for inputs ~~(currently only Android)~~.

Unsure if also clicks should feedback or only long press... What do you think?

### CODE REVIEW

**Comment by IvanSavenko** [comment]
Would be nice to have it toggleable or to use OS API to detect whether user has haptic feedback enabled

**Comment by dydzio0614** [comment]
@Laserlicht probably only long press for now to go safe, though I am not android player so I am not very good at imagining this in practice

**Comment by Laserlicht** [comment]
@IvanSavenko Added setting. Have to test it first. But should be correct?

@dydzio0614 Okay - only Right-"Click" for now. @IvanSavenko I think we should also add it for an action in the radial wheel.

I also added iOS. But I cannot test it. Can someone confirm the functionality?

**Comment by dydzio0614** [comment]
will probably take a while to get iOS test results...

**Comment by Laserlicht** [comment]
> will probably take a while to get iOS test results... 

Maybe I can test with the iPhone from someone...

**Comment by IvanSavenko** [comment]
@Laserlicht 
Check InputSourceTouch::handleEventFingerDown
You need to update settings once they change, similar to long touch duration. Kind of workaround solution at the moment, but works.

@dydzio0614 Might get one eventually (iPhone, not Patreon). My current phone seems to be slowly but surely approaching his end.

**Comment by kambala-decapitator** [comment]
Will try to test on iPad in the next few days

File: client/eventsSDL/InputSourceTouch.cpp [hunk 1]
```diff
@@ -22,6 +22,13 @@
 #include "../gui/MouseButton.h"
 #include "../gui/WindowHandler.h"
 
+#ifdef VCMI_ANDROID
+#include "../../lib/CAndroidVMHelper.h"
+#endif
+#ifdef VCMI_IOS
+#include "../ios/utils.h"
+#endif
```
Comment (kambala-decapitator) [inline client/eventsSDL/InputSourceTouch.cpp:30]:
```suggestion
#if defined(VCMI_ANDROID)
#include "../../lib/CAndroidVMHelper.h"
#elif defined(VCMI_IOS)
#include "../ios/utils.h"
#endif
```

File: client/eventsSDL/InputSourceTouch.cpp [hunk 2]
```diff
@@ -287,3 +297,15 @@ void InputSourceTouch::emitPinchEvent(const SDL_TouchFingerEvent & tfinger)
...
 		GH.events().dispatchGesturePinch(lastTapPosition, distanceNew / distanceOld);
 }
+
+void InputSourceTouch::hapticFeedback() {
+	if(hapticFeedbackEnabled) {
+#if defined(VCMI_ANDROID)
+		CAndroidVMHelper vmHelper;
+		vmHelper.callStaticVoidMethod(CAndroidVMHelper::NATIVE_METHODS_DEFAULT_CLASS, "hapticFeedback");
+#endif
+#if defined(VCMI_IOS)
```
Comment (kambala-decapitator) [inline client/eventsSDL/InputSourceTouch.cpp:307]:
```suggestion
	if(hapticFeedbackEnabled)
	{
#if defined(VCMI_ANDROID)
		CAndroidVMHelper vmHelper;
		vmHelper.callStaticVoidMethod(CAndroidVMHelper::NATIVE_METHODS_DEFAULT_CLASS, "hapticFeedback");
#elif defined(VCMI_IOS)
```

File: client/ios/utils.mm [hunk 3]
```diff
@@ -43,4 +43,11 @@ void hideLoadingIndicator()
 	[indicator removeFromSuperview];
 	indicator = nil;
 }
+
+void hapticFeedback()
+{
+	UIImpactFeedbackGenerator *hapticGen = [[UIImpactFeedbackGenerator alloc] initWithStyle:(UIImpactFeedbackStyleLight)];
+	[hapticGen impactOccurred];
+	hapticGen = NULL;
```
Comment (kambala-decapitator) [inline client/ios/utils.mm:51]:
```suggestion
	auto hapticGen = [[UIImpactFeedbackGenerator alloc] initWithStyle:UIImpactFeedbackStyleLight];
	[hapticGen impactOccurred];
```

**Comment by Laserlicht** [comment]
Tested on android and iPhone & is working

File: config/widgets/settings/generalOptionsTab.json [hunk 4]
```diff
@@ -106,6 +110,12 @@
 					"help": "vcmi.systemOptions.longTouchButton",
 					"callback": "setLongTouchDuration",
 					"created" : "touchscreen"
+				},
+				{
+					"name": "hapticFeedbackCheckbox",
+					"help": "vcmi.systemOptions.hapticFeedbackButton",
+					"callback": "hapticFeedbackChanged",
+					"created" : "touchscreen"
```
Comment (IvanSavenko) [inline config/widgets/settings/generalOptionsTab.json:118]: For consistency - use same condition as in label creation
```suggestion
					"created" : "mobile"
```

File: client/windows/settings/GeneralOptionsTab.cpp [hunk 5]
```diff
@@ -190,6 +194,9 @@ GeneralOptionsTab::GeneralOptionsTab()
 	std::shared_ptr<CToggleButton> framerateCheckbox = widget<CToggleButton>("framerateCheckbox");
 	framerateCheckbox->setSelected(settings["video"]["showfps"].Bool());
 
+	std::shared_ptr<CToggleButton> hapticFeedbackCheckbox = widget<CToggleButton>("hapticFeedbackCheckbox");
+	hapticFeedbackCheckbox->setSelected(settings["general"]["hapticFeedback"].Bool());
```
Comment (IvanSavenko) [inline client/windows/settings/GeneralOptionsTab.cpp:198]: This checkbox might not exist, e.g. on desktop systems
```suggestion
	if (hapticFeedbackCheckbox)
		hapticFeedbackCheckbox->setSelected(settings["general"]["hapticFeedback"].Bool());
```

**Review by IvanSavenko — APPROVED** [review]
Please fix these two issues, but looks OK othervice

File: config/widgets/settings/generalOptionsTab.json (hunk 4 above)
Comment (dydzio0614) [reply config/widgets/settings/generalOptionsTab.json:118]: unrelated but how do we handle stuff like microsoft surface tablets? :P

Comment (IvanSavenko) [reply config/widgets/settings/generalOptionsTab.json:118]: Same as SDL :)

Generally, touch input should work on any system as long as it is supported by SDL. I see some conflicting info on SDL touch input on Windows, so not sure if it is supported.

Linux-based systems (non-Android) should work, e.g. Steam Deck, but obviously not tested.

**Comment by IvanSavenko** [comment]
Also - perhaps we should move autosaves to *end* of turn? Right now I think we do them at turn start, unlike H3

**Comment by Laserlicht** [comment]
@IvanSavenko Done :)

**Comment by IvanSavenko** [comment]
Just noticed one thing - we play haptic feedback not when popup menu appears, but whenever there is a long touch.

What happens if you try to use long touch on areas that don't have popup menu? For example outside of buttons in main menu?

**Comment by IvanSavenko** [comment]
Yeah, probably only when menu opens. So should be moved into `if (GH.windows().isTopWindowPopup())` block few lines below

**Pull request merged**